WalletWasabi/Affiliation/CoinjoinRequestsUpdater.cs

```diff
@@ -0,0 +1,216 @@
+using System.Linq;
+using System.Collections.Immutable;
+using System.Collections.Generic;
+using WalletWasabi.Affiliation.Models;
+using NBitcoin;
+using WalletWasabi.WabiSabi.Backend.Rounds;
+using System.Threading;
+using System.Threading.Tasks;
+using WalletWasabi.WabiSabi.Backend.Events;
+using WalletWasabi.Bases;
+using System.Collections.Concurrent;
+using WalletWasabi.Affiliation.Models.CoinjoinRequest;
+
+namespace WalletWasabi.Affiliation;
+
+public class CoinjoinRequestsUpdater : PeriodicRunner
+{
+	private static readonly TimeSpan AffiliateServerTimeout = TimeSpan.FromSeconds(60);
+	private static readonly TimeSpan Interval = TimeSpan.FromSeconds(1);
+
+	public CoinjoinRequestsUpdater(Arena arena, ImmutableDictionary<AffiliationFlag, AffiliateServerHttpApiClient> clients, CoinJoinRequestRequestsSigner signer)
+		  : base(Interval)
+	{
+		Arena = arena;
+		Clients = clients;
+		Signer = signer;
+
+		CoinjoinRequests = new();
+		RoundData = new();
+		RoundsToUpdate = new();
+
+		AddHandlers(arena);
+	}
+
+	private Arena Arena { get; }
+	private CoinJoinRequestRequestsSigner Signer { get; }
+	private ImmutableDictionary<AffiliationFlag, AffiliateServerHttpApiClient> Clients { get; }
+	private Dictionary<uint256, Dictionary<AffiliationFlag, byte[]>> CoinjoinRequests { get; }
+	private ConcurrentDictionary<uint256, RoundData> RoundData { get; }
+	private Queue<FinalizedRoundDataWithRoundId> RoundsToUpdate { get; }
```

Comment:
After a quite glance i was not able to find any obvious problem with this as it is. However, it is really fragile and it would be better to have lock to make sure we maintain consistence across the multiple data structures.

Reply:
I will consider it. Nevertheless, I'm not really a fan of global locks, since it can lead to unexpected delays or even deadlocks. Which is undesirable as long as `Arena_AffiliationAdded`, `Arena_RoundCreated`, `Arena_CoinjoinTransactionAdded` and `Arena_RoundPhaseChanged` run synchronously (by the way, this can be solved using queues).

I can image that I would have two locks:
  * one for `CoinjoinRequests` (which is a nested `ConcurrentDictionary`) and
  * one for `RoundData` (which is a `ConcurrentDictionary` compounded with `ConcurrentBag`)